lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +319,318 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
-              // subtask as having failed.
-              SubtaskFailed = true;
-            }
-          }
-
-          ExecutingTasks.erase(Pid);
-          FinishedFds.push_back(fd.fd);
-        }
-      } else if (fd.revents & POLLNVAL) {
-        // We passed an invalid fd; this should never happen,
-        // since we always mark fds as finished after calling
-        // Task::finishExecution() (which closes the Task's fd).
-        llvm_unreachable("Asked poll() to watch a closed fd");
-      }
-
-      fd.revents = 0;
-    }
+    Optional<std::vector<int>> FinishedFds =
+        readFromReadyFdsReturningFinishedOnes();
+    if (!FinishedFds)
+      return true;
 
-    // Remove any fds which we've closed from PollFds.
-    for (int fd : FinishedFds) {
-      auto predicate = [&fd] (struct pollfd &i) {
-        return i.fd == fd;
-      };
+    stopPolling(*FinishedFds);
+  }
+  return false;
+}
 
-      auto iter = std::find_if(PollFds.begin(), PollFds.end(), predicate);
-      assert(iter != PollFds.end() && "The finished fd must be in PollFds!");
-      PollFds.erase(iter);
-    }
+bool TaskShepherd::startUpSomeTasks() {
+  while (!TasksToBeExecuted.empty() &&
+         TasksBeingExecuted.size() < MaxNumberOfParallelTasks) {
+    std::unique_ptr<Task> T(TasksToBeExecuted.front().release());
+    TasksToBeExecuted.pop();
+    if (beginExecutingATask(T.get()))
+      return true;
+    startPollingFdsOfTask(T.get());
+    TasksBeingExecuted.add(std::move(T));
+  }
+  return false;
+}
+
+void TaskShepherd::startPollingFdsOfTask(const Task *T) {
+  FdsBeingPolled.push_back({T->getPipe(), POLLIN | POLLPRI | POLLHUP, 0});
+  // We should also poll T->getErrorPipe(), but this introduces timing
+  // issues with shutting down the task after reading getPipe().
+}
+
+TaskShepherd::PollResult TaskShepherd::pollTheFds() {
+  assert(!FdsBeingPolled.empty() &&
+         "We should only call poll() if we have fds to watch!");
+  int ReadyFdCount = poll(FdsBeingPolled.data(), FdsBeingPolled.size(), -1);
+  callBacks.PolledAnFd();
+  if (ReadyFdCount != -1)
+    return PollResult::NoError;
+  return errno == EAGAIN || errno == EINTR ? PollResult::SoftError
+                                           : PollResult::HardError;
+}
+
+bool TaskShepherd::beginExecutingATask(Task *T) {
+  if (T->execute())
+    return true;
+  callBacks.TaskBegan(T->getPid(), T->getContext());
+  return false;
+}
+
+Optional<std::vector<int>>
+TaskShepherd::readFromReadyFdsReturningFinishedOnes() {
+  std::vector<int> finishedFds;
+  for (struct pollfd &fd : FdsBeingPolled) {
+    const int fileDes = fd.fd;
+    const short receivedEvents = fd.revents;
+    fd.revents = 0;
+    Task &T = TasksBeingExecuted.findTaskForFd(fileDes);
+    const bool fdHungUp = reactToFdEvents(receivedEvents, fileDes, T);
+    if (!fdHungUp)
+      continue;
+    finishedFds.push_back(fileDes);
+    const bool hadError = TaskCompleter(T, callBacks.TaskFinished, callBacks.TaskSignalled)
```

Comment:
I can see _some_ motive (in terms of avoiding long argument lists) for having a separate class for `TaskShepherd` but `TaskCompleter` here really just feels like a set of static functions, and constructing a class instance feels like it makes readability worse.

Reply:
Fascinating! I wanted to group all the completion functionality together. 
I do see your point here.
How about static functions in a TaskCompleter class? They don't feel like they have much to do with the TaskShepherd. (I know, there must be a better name for that.)

Reply:
I suppose; but if so, is the class doing more than just serving as (say) a C++ namespace there, for grouping?

Reply:
Check out what I've done with this code! Hope it reads better.